### DISCUSSION TITLE
Call on time integration in Mesa (Tuesday 27 aug 14:30 CEST)

### DISCUSSION THREAD

**Discussion by EwoutH** [mesa/mesa#2247] [General]
We are planning a call Tuesday 27 aug 14:30 CEST (12:30 PM GMT) to discuss how the concept of time should be integrated in Mesa. The call will take place from our General channel on Matrix, as usual.

[![chat](https://img.shields.io/matrix/project-mesa:matrix.org?label=chat&logo=Matrix)](https://matrix.to/#/#project-mesa:matrix.org)

### Proposed agenda

- Who is our target audience?
  - Experience levels
  - Only ABM or also SD / DES / hybrid?
- How would each of these audience(s) best experience time management?
- Are there specific use cases we want to enable for one or more of these specific audiences?
- What specific constraints do we have (technical, engineering capacity, etc.)?
- What would a conceptual approach to time in line with these be?
  - https://github.com/projectmesa/mesa/discussions/2227
  - https://github.com/projectmesa/mesa/discussions/2228
  - https://github.com/projectmesa/mesa/discussions/2229
- How do we integrate with the DEVS simulator
  - https://github.com/projectmesa/mesa/pull/2248

### Write-up of current state of affairs
_Thanks @quaquel for the write-up!_

PR #2223 to automatically increase time and step count in the model touched on many conceptual subjects on time. After more discussion, [5 discussions](https://github.com/projectmesa/mesa/pull/2223#issuecomment-2298570587) where split of.

#### Where consensus is already reached

1. MESA should be easy to use out of the box and offer convenient ways of expanding to more advanced use cases. The current approach where the user just calls model.step seems perfect out of the box.
2. It would be good to have easy support in MESA for tracking the number of steps/time. There are different design directions here. These range from some convenience method/function/annotation that the user can opt into, to automagically tracking it as [the original PR]( #2223) proposed.
3. There needs to be agreed convention for the order of things: incrementing step/time, executing the model step, and collecting data.

#### Where consensus still needs to be reached

1. Is there a use case for allowing the user control over the time/step increment, and, if we agree on the need to support this use case, how to offer this control?
2. Should MESA by default track both step and time. Currently both attributes exist on the model class. But do we want to continue this in MESA 3?
3. What is the order of things. In particular, when to increment step and when to increment time. What are existing conventions for this in other modeling tools and libraries?
4. How to implement point 2. Let it be handled automagically, or force the user to control it explicitly.

### Conclusion of the call
- There is no agreement between the long term usefulness of a continuous time variable built-in. Potential use cases are the simulator, having a datetime and human-interpreble time. For now we keep it as is: `model._time` is (by default) a copy of `model._steps`, as a way to move the discussion forward
- We adopt the count-do-collect order. That's a different schedulers, who first do and then increase the step counter.
- AgentSet operations are generally preferred above schedulers, but we'll keep them around for another 
- Implementation: We're wrapping the step
 
Maintainers, feel free to edit/update this post directly.

**Comment by rht** [comment]
> The current approach where the user just calls model.step seems perfect out of the box.

Not yet fully resolved, https://github.com/projectmesa/mesa/discussions/2227#discussioncomment-10419822. There is an "explicit is better than implicit" argument in this context. Note that NetLogo opted to use an explicit `tick` call.

**Reply by quaquel** [reply]
I think you misunderstand what I was trying to say. In point 1 of agreement, I am solely refering to 

```python

for _ in range(100):
    model.step()
```

NetLogo's tick command is one way of solving the second point in the list of agreements: an explicit function for advancing the model's step counter. It seems we agree that some form of control is desirable, but we don't agree yet on how to implement this.

I have updated the original post to make this more clearly as a fourth item. We agree on the principle (point 2, list of agreements) but don't agree yet on the implementation details (now point 4 of disagreements).

**Reply by EwoutH** [reply]
Yeah, I'm getting some doubts about increasing the time automatically. Maybe `self._advance_time()` wasn't that bad (but we need a prettier name).

I do also still think there are also benefits of doing it automatically, for consistency and (beginner) friendliness. A convention of the order would also help for consistency (but you can't trust only a convention).

<hr>

Since PEP 20 always seems to cover everything, I think we have a few conflicting arguments here:

> Beautiful is better than ugly.

Points toward doing it automatically, it's by far the most elegant.

> Explicit is better than implicit.

Points to letting the user do it.

> There should be one-- and preferably only one --obvious way to do it.

Points towards a fixed order - preferably an implantation over a convention

<hr>

So I think on a higher level, need to determine how much flexibility we're willing to offer on this. I think there are 4 main

1. We fix the order in place, increase step time automatically and don't allow doing it otherwise (you can't disable the internal clock or tick on another time)
2. We fix the order in place, increase step time automatically and allow doing it otherwise (now you can), but don't guarantee to support it in other places (datacollector, viz, etc.)
3. We offer a convention for the order and a method to increase the step/time, which you need to call yourself. We support multiple orders of doing things.
4. We expect you to increase a step/time variable yourself
5. We don't handle time in the model at all.

Personally I'm nudging towards 2, but I also see the perspective of 3.

**Reply by quaquel** [reply]
My idea is in line with what was mentioned by @tpike3 in some other place: have good defaults and allow for extension/modification for advanced use cases. I think this straddles the fence of pep20 nicely. 

For me, a good default would be automatic counting of the number of times model.step is called and as argued by @corvince in the context of the future of datacollection by default collect all simulation state data over time. Counting steps is such a common use case that it should just be there in my opinion. Although I am the developer of pynetlogo, I have never built a netlogo model so I am stupified to find out that you have to explicitly call tick in netlogo to advance the step/tick counter. 

Where I struggle is how to cover the extension/modification part. First, I fail to see use cases for messing with a simple step counter. Even the staged activation increments make little senste to me. Since I don't know of use cases, I have no idea what to design for. If you want to track the intermediate steps between stages, why not let a user code this up in whatever way they see fit?

The proposed solution in #2223 through keyword arguments I really don't like. It burderns the step method with a buch of kwargs that I don't see users use very often, thus making the documentation for one of the key methods that users must implement in their model more complicated then needed. It also limits flexibility because one or two keywords have now become reserved keywords. One solution that might be more pallatable to me (but see point 1 - I failt to see the use case), would be to have a simple step_delta (better name needed) attribute on the model. This is set to 1 and if you want to change the behavior of the step counter you just change this attribute. 

Third, I do see use cases for advanced management of time (in contrast to steps). But this raises many issues about datetime and linking ticks in an ABM to real world time. Some of the experimental functionality of the two simulator classes would be a potential entry point for more advanced time management. Still, it seems that advanced time management would require some kind of clock class and raises other issues surrounding not just just time, but date management, and calenders. I personally atm am inclinded to simply remove time from the model class altogether. This further simplifies the default API.

**Reply by EwoutH** [reply]
For some context, the default [example](https://ccl.northwestern.edu/netlogo/docs/programming.html#tick-counter) NetLogo gives is:
```
to go
  ask turtles [ fd 1 ]
  tick
end
```
They also allow fractional ticks, with a built-in command [`tick-advance`](https://ccl.northwestern.edu/netlogo/docs/dictionary.html#tick-advance).

**Reply by DrEntropy** [reply]
>Although I am the developer of pynetlogo, I have never built a netlogo model so I am stupified to find out that you have to explicitly call tick in netlogo to advance the step/tick counter.

In netlogo,  `go` (the equivalent of `step`) is pure convention.  You can call this whatever you want.  So there is no way for the system to know that a step (tick) has happened without telling it with `tick`.  

Mesa is different in this respect,  so it seems fine to me that Mesa could use calls to `step` to know it should advance a step counter.   However,  I am also good with having to explicitly call some kind of tick function, or even having to increment `steps` manually.  Just as long as there is some convention and it is well documented.   Any of these options seem less opaque to me than the current situation where the scheduler takes care of it in (smuggled in with its own wrapped_step). 

In any event I look forward to reading about the results of your deliberations!

**Reply by quaquel** [reply]
> In netlogo,  go (the equivalent of step) is pure convention. You can call this whatever you want. So there is no way for the system to know that a step (tick) has happened without telling it with tick.

Thanks for the further clarification

**Reply by Corvince** [reply]
> (smuggled in with its own wrapped_step).

Now that you say that, that brings me to another idea. We could also make this more explicit instead of magic by providing a function wrapper. So users would do something like

```python
@mesa.step
def step(self):
    ...
```

This would make it obvious that something is happening under the hood. For example we could only then assign "model.steps", so it doesn't have to be a reserved keyword. And we would be able to accommodate more things in the future, for example add automatic time advancement via a flag (so for example `@mesa.step(time=True)`. Also you cold name model.step differently or have muiltpile functions that increase the step, just all wrapped with `@mesa.step` (so it becomes more extensible).

**Reply by EwoutH** [reply]
The more we’re discussing it the more I’m leaning to just letting users do it themselves. It’s not as clean as I would like, but sometimes explicit is better than implicit.

Thanks for your input, @DrEntropy! If you would like to join the call today, please do!

**Reply by EwoutH** [reply]
From a historical perspective, we’ve always had schedulers that increased the `scheduler.steps` counter automatically. Maybe that’s where this is originating from.

**Reply by quaquel** [reply]
> The more we’re discussing it the more I’m leaning to just letting users do it themselves. It’s not as clean as I would like, but sometimes explicit is better than implicit.

My take away is actually the opposit. We need to figure out a good default and a way of making it easy to extend. @EwoutH bolder analogy made elsewhere applied here. I actually think it might be possible to use annotations (as suggested by @Corvince below), if the user does not annotate their model.step, MESA automagically adds a default annotation that does increment step, execute user's step method, and call datacollect.collect. Then we also capture this order explicitly. If users want more control, their is a set of annotations, possibly with keyword arguments,to further control and refine how mesa behaves.

**Reply by EwoutH** [reply]
I’m not strictly against it, but I find annotations quite a difficult concept for starting modelers.

**Edit:** We can users just link [this video](https://youtu.be/BE-L7xu8pO4).

**Reply by Corvince** [reply]
Oh I like the suggestion by @quaquel to have opt-out/modify annotations! That's beginner friendly, but highly customizable.

**Reply by EwoutH** [reply]
Jan and I discussed this a little further, and I can get behind the step-wrapping-decorator-with-keyword-arguments. It's basically the same kind of wrapping I did in https://github.com/projectmesa/mesa/pull/2223, but then with a decorator with keyword arguments instead of keyword arguments in the subclassed step.

**Comment by quaquel** [comment]
As input to the conversation, it might be useful to take a look at [chapter 3 from Zieg'ers Theory of Modeling and Simulation](https://www.sciencedirect.com/science/article/pii/B9780128133705000110). Note that this chapter informed the design of the current experimental ABMSimulator and DEVSSimulator.

**Reply by rht** [reply]
(Reporting that the PDF can't be found in any Library Genesis mirrors.)

An LLM summary (of the book) said
> General Flow in DEVS:
> - Advance the Clock: The simulation clock is advanced to the time of the next event (whether internal or external).
> - State Update: After the clock is set to the correct time, the state of the system is updated based on the event.

This makes sense in the context of DEVS, because the new state at time `Aug 27 2024 + 1000 days` should computed based on the newly advanced time, not the previous time (Aug 27), which would be relatively a long time ago in the past. The analogy of the time jump with ODE solving would be to use the adaptive timestep. I can only say that DEVS maps to backward diff solving of ODE. But ODE solving would allow both forward and backward update.

I tried looking at the DifferentialEquations.jl code, but couldn't get through it (the abstractions are so thick), except for the algorithms specifically using [backward differentiation](https://github.com/search?q=repo%3ASciML%2FOrdinaryDiffEq.jl+backward+diff&type=code).

https://docs.sciml.ai/DiffEqDocs/latest/basics/integrator/#Initialization-and-Stepping

**Reply by rht** [reply]
![2024-09-01T03:47:47,459759444-04:00](https://github.com/user-attachments/assets/d2f807cb-c705-46cb-ba44-e46d7ba34e83)
This is the algorithm as shown in the chapter 3 of the book. There is no explicit increment of t, but I'd assume t starts with Ti, then the `while` check happens, then y(t) and q(t+1) are calculated, t is incremented, then the next while check happens, and so on. This is equivalent to the forward algorithms, but also could be adapted to be backward (q(t+1) = delta(q(t), x(t+1))), as the book later discusses about forward and backward Euler.

But this would mean that:
1. q(0) is the initial state
2. initial output trajectory, y(0), is calculated from q(0) and x(0)
3. the next state, q(1) is calculated from Ti, not Ti+1 (Ti is the initial time)
4. There is no notion of `steps` as in Mesa, but rather a more general t, that once again, I think is incremented last

In Mesa syntax, (my interpretation of) Zeigler's convention would be
```python
# Model.__init__
state_0 = prepare()
steps = 0

# Step 0, the first model step
y_0 = datacollector.collect()
state_1 = agentset.step()  # uses state_0 inside
steps += 1

# Step 1
y_1 = datacollector.collect()
state_2 = agentset.step()  # uses state_1 inside
steps += 1
```

The differences with the current Mesa:
1. datacollector is called first
2. datacollector is called only inside the model step, but never in `Model.__init__()`

**Reply by rht** [reply]
The current Mesa would be
```python
# Step 0, i.e. defined in __init__, there is not a single step() call
steps = 0
state_0 = prepare()
y_0 = datacollector.collect()

# Step 1, i.e. the first model step
steps += 1
state_1 = agentset.step()  # uses state_0 inside
y_1 = datacollector.collect()
```

**Reply by quaquel** [reply]
It is indeed annoying that the example does not tell you where $t$ is incremented.

But what if, while in the step method, a user wants to collect data? What is the "time index" assigned to this data point? In my view, the moment you are inside the state update (i.e. `model.step` in MESA), you are no longer in $t_0$. So, I don't see any other option then to first increment $t$ before executing the state updates associated with this new $t$. This, of course, assumes that $t$ is just the "time index".

**Reply by rht** [reply]
I consider the $y(t)$ (the output) to encompass what the data collector does. The problem with Zeigler's scheme is that at a given step, there are y(t) and q(t+1) that correspond to a different time. Mesa's current scheme is more consistent with the step, state, output/data collection all in sync.

**Reply by EwoutH** [reply]
Sometimes it's worth asking the question: "Is this the best use of time to improve mesa?"

Two totally unrelated points:

- Making and updating [examples](https://github.com/projectmesa/mesa-examples) let's you find out what's actually working and what's not
- Go talk to students and modellers, what Mesa part did you find most difficult to understand or what did you spend most time on to get working?

But I guess academics is also about debate, so whatever

**Comment by EwoutH** [comment]
I started a discussion on how the DEVS simulator could integrate in this picture:

- https://github.com/projectmesa/mesa/pull/2248

**Comment by Corvince** [comment]
Since unfortunately I wont be able to make it to the call, here is my opinion on the matter.

First of all I want to repeat that I now think we should not conflate the steps counter and time too much. This is somewhat already present in that we have currently have both steps and time, but also in the discussion "steps" is sometimes regarded as a time variable. And while I think this is a valid approach for most models, I think we should just conceptually define it simply as what it is - the number of times "model.step" is called. And we define this as a significant change in model state - hence we trigger visualization updates in response to this event and we collect the number of steps whenever the datacollector collects data. 
From this definition it also follows that disabling it or changing the step size doesn't really make any sense. If you really really don't want to use it - just don't call model.steps, but move your model logic to another function. There is no need to use a model.step, just a convention. And we can expand that convention to also count the number of times it has been called. 

But of course it also makes sense for some models to have a more fine granular handling of time - and I am very excited for which features you will discuss could be added to a possible time or clock function. But I would see this as a completely additional module and I don't think we need to hard code something into our model base class. Just let users add it by declaring model.time = AwesomeMesaClock()

Lastly I want to add my opinion on when to call model.step. I had the impression that almost everyone suggested doing it at the beginning, before user actions. I agree with this opinion. It's less clear cut for time, but this is a separate issue.

**Comment by EwoutH** [comment]
Without judgement, here’s an example of someone wanting to increase the step counter by some other amount:

- https://github.com/projectmesa/mesa-examples/pull/138

**Reply by EwoutH** [reply]
@vitorfrois you’re very welcome to join the call by the way!

**Reply by quaquel** [reply]
I looked at the code. In my view, this hijacks _steps to count the number of agent activations. It would be cleaner to just at a seperate agent_activations attribute and increment that after each single agent activation.

**Comment by EwoutH** [comment]
**Conclusions from the meeting**
- There is no agreement between the long term usefulness of a continuous time variable built-in. Potential use cases are the simulator, having a datetime and human-interpreble time. For now we keep it as is: `model._time` is (by default) a copy of `model._steps`, as a way to move the discussion forward
- We adopt the count-do-collect order. That's a different from some of the existing schedulers (e.g., staged activation), who first do and then increase the step counter.
- AgentSet operations are generally preferred above schedulers, but we'll keep them around for another 
- Implementation: We're wrapping the step
